20250226-actors-pubsub.md

```diff
@@ -0,0 +1,158 @@
+}
+
+// PublishActorEventRequestAlpha1 is the message to publish event data to an
+// actor ID.
+message PublishActorEventRequestAlpha1 {
+  // actor_type is the type of the actor to publish to.
+  // Required.
+  string actor_type  = 1;
+
+  // actor_id is the ID of the actor_type to publish to.
+  // Required.
+  string actor_id = 2;
+
+  // method is the endpoint of the actor to invoke with the published data.
+  // Required.
+  string method = 3;
+
+  // The data which will be published to topic.
+  bytes data = 4;
+
+  // The content type for the data (optional).
+  optional string data_content_type = 5;
+
+  // The metadata passing to pub components
+  //
+  // metadata property:
+  // - key : the key of the message.
+  optional map<string, string> metadata = 6;
+}
+```
+
+### Event Message Routing
+
+Upon the Daprd runtime receiving an actor PubSub message, the runtime will wrap the message with a CloudEvent envelope as usual.
+The PubSub used will be the static PubSub Component marked as the actor PubSub, the topic string built using the format above.
+If no actor PubSub is defined, an appropriate typed error will be returned to the client.
+
+The Actor PubSub message CloudEvent envelope will have the following fields added to allow for routing of the message on delivery.
+
+```
+dapr.actors.id
+dapr.actors.method
+```
+
+An example of the CloudEvent envelope:
+
+```json
+{
+  "specversion": "1.0",
+  "type": "com.dapr.event.sent",
+  "source": "joshvanl",
+  "id": "5929aaac-a5e2-4ca1-859c-edfe73f11565",
+  "time": "1970-01-01T00:00:00Z",
+  "datacontenttype": "application/json",
+  "data": {
+    "message": "Hello, World!"
+  },
+  "dapr.actors.id": "MyActorID",
+  "dapr.actors.method": "MyActorMethod"
+}
+```
+
+Upon the daprd runtime receiving an actor PubSub message, the runtime will unwrap the CloudEvent envelope and route the message to the appropriate actor instance.
+The actor type and ID to route is easily extracted from the topic the message was received from, and the ID from the CloudEvent envelope.
+The data payload will be serialized just as is for normal subscriptions today.
+The method field will be used to invoke the actor at that method with the data payload.
+The result of that invocation will determine the response code sent back to the PubSub broker component.
+
+All Daprds who share a hosted actor type will subscribe to the same PubSub topic.
+These Daprds will receive messages from the subscription for that type, destined for IDs which they often will not be currently hosting.
+These messages will be routed (proxied) to the correct daprd host, based on the placement table hash.
+Proxying requests like this is already implemented and used in the actor runtime engine today, so no work needs to be added to support this.
+
+SDKs will need to be updated to support publishing actor PubSub messages.
+No changes to SDKs need to be made to support _receiving_ actor PubSub messages.
```

Comment:
I don't know about this - at least .NET would need to know about the start/end of the turn for reentrancy and state cache purposes.